Rest api dev

tlatmsrud opened the pull request.

hannut91 committed:
웹 의존성 버전을 업그레이드 하라

의존성 버전으로 인해 실행이 안되는 문제가 있어서 최신화 합니다.

hannut91 committed:
Merge pull request #82 from CodeSoom/update-version

웹 의존성 버전을 업그레이드 하라

tlatmsrud committed:
price가 문자열로 들어왔을 때의 Vaildation 및 예외처리

tlatmsrud committed:
필수값 누락에 대한 예외문구 처리

tlatmsrud committed:
물품 수정 API에 대한 Vaildation 예외처리 및 문구 테스트

tlatmsrud committed:
ProductData에 대한 Vaildation 및 예외문구 추가

tlatmsrud committed:
ProductData에 대한 Vaildation 테스트

tlatmsrud committed:
ProductData 에 대한 Vaildation Test 로직 수정

테스트 실패로 인한 수정

tlatmsrud committed:
NotFoundErrorAdvice를 GlobalErrorAdvice로 변경

Vaildation 예외 Handler 기능 추가로 인한 클래스명 변경

tlatmsrud committed:
GlobalErrorAdvice 에서 GlobalExceptionHandler로 이름 변경

tlatmsrud committed:
ProductData에 대한 DozerMapper 적용 및 Service, Test, DTO 수정

1. ProductService에 ProductData > Product 로 Mapping하는 로직 추가
2. ProductServiceTest에 DozerMapper Bean에 대한 생성자 DI 부분 추가
3. ProductData에 대한 Mapping 정보 추가

tlatmsrud committed:
DozerMapper 의존성 추가

tlatmsrud committed:
User에 대한 Controller, Service, Repository, DTO, Entity 생성

tlatmsrud committed:
테스트 실패로 인한 로직수정

tlatmsrud commented:
e2e 테스트 도중 요구사항 관련하여 여쭤볼게 있습니다.
1. 등록시에는 이메일이 필수값이고, 수정시에는 필수가 아닌게 맞나요?
2. without required parameter e2e 테스트에서 계속 201 로 응답이 오고 있습니다. 실 테스트에는 400으로 떨어지는데 혹시 user.spec을 보니 36번째줄에 파라미터로 user()를 넣던데 테스트 데이터 문제인건가요??

hannut91 commented:
> 등록시에는 이메일이 필수값이고, 수정시에는 필수가 아닌게 맞나요?

맞습니다. 이메일은 변경할 수 없도록 되어있어서 필수로 받지 않습니다.

hannut91 commented:
> without required parameter e2e 테스트에서 계속 201 로 응답이 오고 있습니다. 실 테스트에는 400으로 떨어지는데 혹시 user.spec을 보니 36번째줄에 파라미터로 user()를 넣던데 테스트 데이터 문제인건가요??

자바스크립트 Spread Syntax에 대해서 아셔야 합니다.

```js
{
  ...user(),
  ...it,
}
```

처럼 쓰게되면, `user()`로 만든 데이터에서 주어진 값으로 덮어쓰게 됩니다.

```js
{
  email: `${new Date().getTime()}@test.com`,
  name: 'testuser',
  password: 'password',
  password: '',
}
```

password로 예를들면 위와 같이 작성한 것과 같은 효과이고, key값이 같은 경우 뒤에 작성한 것으로 덮어쓰게 되기 때문에 결국 패스워드가 빈 값으로 요청하고 있습니다.

물론 테스트가 잘못될수도 있으니 자세히 들여다 봐야할 것 같습니다.

tlatmsrud committed:
테스트 커버리지를 맞추기 위한 작업

1. domain, dto 에 대한 테스트 코드 추가
2. 사용하지 않는 로직 및 어노테이션 제거

tlatmsrud committed:
유저 정보 수정 시 필수값 조건 제거

유저정보 수정 요청 시 모든 값이 필수값이 아닌 관계로 Valid를 제거하였음.

tlatmsrud committed:
Validation에 대한 Group 분리 및 적용

1. 등록 및 수정 시 하나의 DTO 에 적용되는 Validation을 분리하기 위해
group 인터페이스를 등록.
2. 유저 생성 그룹 새성 : UserCreateValidationGroup
3. 유저 수정 그룹 생성 : UserUpdateValidationGroup
4. UserData의 Validation 어노테이션에 대한 Grouping 적용
5. UserController의 요청 값에 대한 Validation 그룹 정의

tlatmsrud committed:
Update 테스트에 대한 테스트 데이터에 email 값 제거

이메일은 수정이 안되는 요건에 의해 테스트 요청 값에서 email을 제거함

tlatmsrud commented:
등록, 수정에 대한 요건분석이 잘 되지 않아 많이 헤맸습니다. user.spec에 정의된 요청 값들을 보니 문제가 되는 부분들을 찾을 수 있었고 로컬에서의 테스트는 모두 성공했습니다.
이번 과제를 하며 궁금한 점 몇개가 있었습니다.
1. Validation Group 사용.
 - 동일한 DTO를 사용하여 POST, FETCH 작업을 수행하고 있으며, 작업마다 Validation 조건이 달라 ValidationGroup을 사용하여 처리하였습니다. 허나 이를 적용하기 위해 ValidationGroup 인터페이스를 2개 생성, DTO의 Validation 관련 어노테이션마다 groups 속성 기입, Controller Method 요청 Command 객체에 ValidationGroup 설정 작업을 진행하였습니다. 저는 이게 조금 복잡하다고 생각됩니다. 찾아보니 위 방법과 inner class를 통해 Create, Update 클래스를 각각 만들어 Validation을 분리하는 방법이 있는데 이것 말고 다른 좋은 방법이 있는지 궁금하며, 주로 어떤 방식을 선호하시는지 궁금합니다.

2. 테스트 커버리지를 맞추기 위한 추가작업
 - 최초 Controller, Service에 대한 테스트 코드를 작성 후 테스트 커버리지를 확인하였습니다. 그 결과 롬복 어노테이션으로 인해 Builder의 ToString, DTO의 Setter 메서드 등 실제 테스트 로직을 실행하며 사용되지 않는 메서드들이 주루룩 나오더라구요. 테스트 커버리지 100퍼센트가 주는 의미가 굉장히 크다고 생각하여 이를 맞춰주기 위해 DTO나 domain 객체에 대한 테스트 클래스도 따로 생성하여 커버리지를 맞춰주었습니다.
 이 작업을 통해 의미있는 100%의 커버리지 맞추려면 구현한 모든 클래스(DTO, Domain 포함) 대해 각각의 테스트 클래스를 만들어 검증하는게 좋은지 궁금합니다. 예를들어 Service 테스트 클래스의 로직에서 DTO의 모든 메서드를 사용하게 될 경우 테스트 커버리지가 맞춰지기에 굳이 DTO 클래스에 대한 테스트클래스를 작성할 필요가 없지만, 그래도 DTO 테스트 클래스를 만들어 검증하는게 맞는지에 대한 궁금증입니다!

hannut91 commented:
> 테스트 커버리지를 맞추기 위한 추가작업

커버리지를 100%를 해야한다고 했던 것은 커버리지를 보는 습관을 들이기 위해서 입니다. 억지로 맞추실 필요없습니다. DTO같이 필요없는 것은 커버리지에 제외시키는게 좋을 것 같습니다.

추가로 테스트 커버리지는 내가 혹시 테스트를 빠트리지 않았나 검사하는 용도로 사용하는 것이지 테스트의 품질을 평가하는 것이 아니기 때문에 주의하시는게 좋습니다. 아래 글을 한 번 읽어보세요!

See also
  - https://martinfowler.com/bliki/TestCoverage.html

hannut91 commented:
> Validation Group 사용

같은 DTO를 사용하는데, Validation이 달라지니, 다르게 동작하기 위해서 일종의 분기를 만드신 것 같습니다. 일종의 flag를 이용해서 분기를 만든 것이고 이건 기능적 응집도라고 하는 응집성이 생긴 것입니다.

저라면 완전히 같더라도 다른 DTO를 만들 것 같습니다. 만약 공통으로 사용하게 되면 두 개를 만족시키기 위해서 한쪽에서만 사용하는 기능들이 들어가고 그러다보면 DTO가 여러가지 일을하면서 거대해질 가능성이 높습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,112 @@
+
+    }
+
+    @Test
+    void updateUserWithExistedId() {
+        UserData userData = UserData.builder()
+                .name("앙김홍집")
+                .email("hongzip@naver.com")
+                .password("123123")
+                .build();
+
+        User user = userService.updateUser(DEFAULT_ID, userData);
+
+        assertThat(user.getId()).isEqualTo(DEFAULT_ID);
+        assertThat(user.getName()).isEqualTo("앙김홍집");
+    }
+
+    @Test
+    void updateUserWithNotExistedId() {
+        UserData userData = UserData.builder()
+                .name("앙김홍집")
+                .email("hongzip@naver.com")
+                .password("123123")
+                .build();
+
+        assertThatThrownBy(() -> userService.updateUser(INVALID_ID, userData))
+                .isInstanceOf(UserNotFoundException.class);
+    }
```
BDD형식의 테스트를 시도해보면 좋을 것 같습니다. BDD는 주어진 상황이 달라지는 경우를 명시적으로 드러낼 수 있습니다.

```java
@Nested
@DisplayName("update")
class Describe_updateUser {
  UserData userData = UserData.builder()
                .name("앙김홍집")
                .email("hongzip@naver.com")
                .password("123123")
                .build();
  
  Long id;

  @Nested
  @DisplayName("사용자가 존재하면")
  class Context_withExistedId {
    @BeforeEach
    void setUp() {
      id = DEFAULT_ID;
    }

    @Test
    @DisplayName("수정한 사용자를 반환한다")
    void updatesUserInformation() {
        User user = userService.updateUser(id, userData);

        assertThat(user.getName()).isEqualTo("앙김홍집");
    }
  }

  @Nested
  @DisplayName("사용자가 존재하지 않으면")
  class Context_with_not_existed_id {
    @BeforeEach
    void setUp() {
      id = INVALID_ID;
    }

    @Test
    @DisplayName("UserNotFoundException예외를 던진다")
    void throwsUserNotFoundException() {
        User user = userService.updateUser(id, userData);

        assertThatThrownBy(() -> userService.updateUser(id, userData))
                .isInstanceOf(UserNotFoundException.class);
    }
  }
}
```

See also
  - https://johngrib.github.io/wiki/junit5-nested/

hannut91 replied:
작성하고보니 엄청나게 티가 나진 않네요. 테스트 케이스가 너무 단순해서

tlatmsrud replied:
넵 그래도 말씀해주신 UserService에 대해 BDD로 수정해보았습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,112 @@
+
+    @Test
+    void updateUserWithExistedId() {
+        UserData userData = UserData.builder()
+                .name("앙김홍집")
+                .email("hongzip@naver.com")
+                .password("123123")
+                .build();
```
테스트용 데이터가 계속 반복해서 쓰이는 것 같네요. 미리 데이터를 만들어놓고 테스트에서 활용하는 것은 어떨까요? 이런 데이터들을 fixture라고 부릅니다

tlatmsrud replied:
fixture를 Enum으로 관리해보았습니다!

tlatmsrud commented:
> > 테스트 커버리지를 맞추기 위한 추가작업

감사합니다. 테스트 커버리지는 테스트되지 않은 부분을 확인하기 위한것이 궁극적인 목적이었군요! 감사합니다.

tlatmsrud committed:
반복되는 고정 테스트 데이터인 Fixture는 Enum 으로 관리하라

tlatmsrud committed:
Fixture Enum 및 테스트코드 리팩토링

tlatmsrud committed:
@Nested를 사용하여 BDD형식의 테스트코드를 구현하라

Service 클래스에 대해 사용자의 행위를 예상한 테스트코드를 작성하라

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -0,0 +1,48 @@
+        return userService.updateUser(id, userData);
+    }
+
+    @DeleteMapping("/{id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void delete(@PathVariable Long id){
+        userService.deleteUser(id);
+    }
```
왜 204로 응답해야 하는지 다음 문서를 읽어봅시다 
> A successful response SHOULD be 200 (OK) if the response includes an
entity describing the status, 202 (Accepted) if the action has not
yet been enacted, or 204 (No Content) if the action has been enacted
but the response does not include an entity.

https://tools.ietf.org/html/rfc2616#section-9.7

tlatmsrud replied:
서버에서 요청은 정상적으로 처리됐으나 응답값이 없을 경우 204를 리턴하는걸로 알고있습니다! 2주차였던가에 피드백해주신 내용이었습니다 ㅎㅎ

hannut91 replied:
아 그랬었군요 ㅎㅎ

hannut91 commented on app/src/test/java/com/codesoom/assignment/dto/UserDataTest.java:
```diff
@@ -0,0 +1,33 @@
+                .password("password")
+                .build();
+
+        String userDataToString = UserData.builder()
+                .id(1L)
+                .name("name")
+                .email("email@naver.com")
+                .password("password")
+                .toString();
```
커버리지 때문에 호출하고 있는 것처럼 보이네요 ㅎㅎ 이런 테스트들은 커버리지에서 제외시키는게 좋아보입니다

tlatmsrud replied:
넵 dto에 대한 테스트 클래스를 모두 제거하였습니다

hannut91 commented on app/src/test/java/com/codesoom/assignment/fixture/UserFixture.java:
```diff
@@ -0,0 +1,27 @@
+import com.codesoom.assignment.domain.User;
+import com.codesoom.assignment.dto.UserData;
+
+public enum UserFixture {
+
+    UPDATE_USER(1L, "앙김홍집","hongzip@naver.com","123123"),
+    CREATE_USER(2L, "앙생성집","sangzip@naver.com","12345");
+
+    private final Long id;
+    private final String name;
+    private final String email;
+    private final String password;
+
+    private UserFixture(Long id, String name, String email, String password){
+        this.id = id;
+        this.name = name;
+        this.email = email;
+        this.password = password;
+    }
+
+    public UserData getUserData(){
+        return new UserData(id, name, email, password);
+    }
+
+}
```
Enum클래스로 테스트용 데이터를 만드셨군요 ㅎㅎ 좋습니다. 나중에 점점 복잡한 케이스가 나온다면 점점 이 부분도 복잡해질겁니다. 그때가 되면 미리 만들어놓는다기보다 쉽게 만들 수 있는 메서드 같은 것을 도입해야 될 수도 있습니다.

tlatmsrud replied:
넵 고민해보겠습니다!

tlatmsrud committed:
DTO에 대한 테스트 클래스를 제거하라

테스트커버리지를 맞추기 위한 불필요 테스트코드이다.
